Update the Lisk Service Configuration Reference page (#1633)
LiskHQ/lisk-service: docs/antora/modules/ROOT/pages/configuration/index.adoc

      log_date_format: 'YYYY-MM-DD HH:mm:ss SSS',
      watch: false,
      kill_timeout: 10000,
      max_memory_restart: '150M',
      autorestart: true,
      env: {
        // --- Remember to set the properties below
        SERVICE_BROKER: 'redis://localhost:6379/0',
        SERVICE_APP_REGISTRY_MYSQL: 'mysql://lisk:password@localhost:3306/lisk',
        ENABLE_REBUILD_INDEX_AT_INIT: 'false',
      },
    },
    {
      name: 'lisk-service-blockchain-connector',
      script: 'app.js',
      cwd: './services/blockchain-connector',
      pid_file: './pids/service_blockchain_connector.pid',
      out_file: './logs/service_blockchain_connector.log',
      error_file: './logs/service_blockchain_connector.err',
      log_date_format: 'YYYY-MM-DD HH:mm:ss SSS',
      watch: false,
      kill_timeout: 10000,
      max_memory_restart: '150M',
      autorestart: true,
      env: {
        // --- Remember to set the properties below
        SERVICE_BROKER: 'redis://localhost:6379/0',
        LISK_APP_WS: 'ws://localhost:7887',
        GEOIP_JSON: 'https://geoip.lisk.com/json',
        USE_LISK_IPC_CLIENT: 'true',
        // LISK_APP_DATA_PATH: '~/.lisk/lisk-core',

>         // ENABLE_TESTING_MODE: 'false',

Reviewer: ```suggestion
```

        // EXCHANGERATESAPI_IO_API_KEY: ''
      },
    },
    {
      name: 'lisk-service-export',
      script: 'app.js',
      cwd: './services/export',
      pid_file: './pids/service_export.pid',
      out_file: './logs/service_export.log',
      error_file: './logs/service_export.err',
      log_date_format: 'YYYY-MM-DD HH:mm:ss SSS',
      watch: false,
      kill_timeout: 10000,
      max_memory_restart: '300M',
      autorestart: true,
      env: {
        SERVICE_BROKER: 'redis://localhost:6379/0',
        SERVICE_EXPORT_REDIS: 'redis://localhost:6379/3',
        SERVICE_EXPORT_REDIS_VOLATILE: 'redis://localhost:6379/4',
      },
    },
  ],
};
----
====

== Microservice Configurations
In this section, we discuss the common and specific configurations of all the microservices of Lisk Service.

=== Common Configurations
The common configurations described in this section can be added to any microservice.
The value of each configuration defined below should be the same across all microservices.

[tabs]
=====
Properties::
+
--
[cols="2,1,1,3,3",options="header",stripes="hover"]
|===
|Property's Name
|Datatype
|Configurable
|Description
|Sample

> | `SERVICE_BROKER`

Reviewer: Below `SERVICE_BROKER`, you could also specify `SERVICE_BROKER_TIMEOUT`.

Author: Please share the following for the `SERVICE_BROKER_TIMEOUT`

- Datatype
- Description
- Sample

Reviewer: Datatype: Number
Description: Inter-microservice request timeout in seconds
Sample: 10

        PORT: "9901",
        HOST: "127.0.0.1",
        // --- Remember to set the properties below
        SERVICE_BROKER: "redis://localhost:6379/0",
        SERVICE_GATEWAY_REDIS_VOLATILE: "redis://localhost:6379/5",
        ENABLE_HTTP_API: "http-status,http-version3,http-exports",
        ENABLE_WS_API: "blockchain,rpc-v3",
        GATEWAY_DEPENDENCIES: "indexer,connector",
        WS_RATE_LIMIT_ENABLE: "false",
        WS_RATE_LIMIT_CONNECTIONS: 5,
        WS_RATE_LIMIT_DURATION: 1, // in seconds
        ENABLE_REQUEST_CACHING: "true",
        JSON_RPC_STRICT_MODE: "false",
        HTTP_RATE_LIMIT_ENABLE: "false",
        HTTP_RATE_LIMIT_CONNECTIONS: 200,
        HTTP_RATE_LIMIT_WINDOW: 10, // in seconds
        HTTP_CACHE_CONTROL_DIRECTIVES: "public, max-age=10",
        ENABLE_HTTP_CACHE_CONTROL: "true",
      },
    },
    // Other types of configurations.
  ],
};
----
--
=====

=== Connector

[tabs]
=====
Properties::
+
--
[cols="2,1,1,3,3",options="header",stripes="hover"]
|===
|Property's Name
|Datatype
|Configurable
|Description
|Sample

| `SERVICE_BROKER`
| string
| ✔
| URL of the microservice message broker (NATS or Redis).
| redis://localhost:6379/0
 

> | `LISK_APP_HTTP`
> | string
> | ✔
> | URL to connect with the Lisk SDK-based application node over HTTP.
> | http://127.0.0.1:7887
>  

Reviewer: Please remove this since we do not use HTTP connections to connect with the blockchain node anymore.
```suggestion
```

| `LISK_APP_WS`
| string
| ✔
| URL to connect with the Lisk SDK-based application node over WebSocket.
| ws://localhost:7887

> | `GEOIP_JSON`

Reviewer: Move it below `LISK_APP_DATA_PATH`.

| string
| ✔
| URL of the GeoIP server
| https://geoip.lisk.com/json

| `USE_LISK_IPC_CLIENT`
| string
| ✔

> | Boolean flag to enable IPC-based connection to the Lisk SDK-based application node.

Reviewer: ```suggestion
| Boolean flag to enable IPC-based connection to the Lisk SDK-based application node. Requires `LISK_APP_DATA_PATH` to be set.
```

| true 

| `GENESIS_BLOCK_URL`
| string
| ✔
| URL of the Lisk SDK-based application's genesis block.
Only to be used when the genesis block is large enough to be transmitted over API calls within the timeout.
| 

> | `LISK_APP_DATA_PATH`

Reviewer: Move it below `USE_LISK_IPC_CLIENT`.

| string
| ✔
| Data path to connect with the Lisk SDK-based application node over IPC.
| ~/.lisk/lisk-core

> | `ENABLE_TESTING_MODE`
> | string
> | ✗
> | 
> | false
> |===
> 

Reviewer: Please remove this. It is only necessary for our CI activities.
```suggestion
```

--
Sample::
+
--
[source,js]
----
module.exports = {
  apps: [
    // Other types of configurations.
    {
      name: "lisk-service-blockchain-connector",
      // [...]
      env: {
        // --- Remember to set the properties below
        SERVICE_BROKER: "redis://localhost:6379/0",
        LISK_APP_HTTP: "http://127.0.0.1:7887",
        LISK_APP_WS: "ws://localhost:7887",
        GEOIP_JSON: "https://geoip.lisk.com/json",
        USE_LISK_IPC_CLIENT: "true",
        GENESIS_BLOCK_URL: "",
        LISK_APP_DATA_PATH: "~/.lisk/lisk-core",

>         // ENABLE_TESTING_MODE: 'false',

Reviewer: ```suggestion
```

      },
    },
    // Other types of configurations.
  ],
};
----
--
=====

=== Coordinator

[tabs]
=====
Properties::
+
--
[cols="2,1,1,3,3",options="header",stripes="hover"]
|===

> |Property's Name
> |Datatype
> |Configurable

Reviewer: ```suggestion
|Environment variable
|Datatype
```

|Description
|Sample

| `SERVICE_BROKER`
| string
| ✔
| URL of the microservice message broker (NATS or Redis).
| redis://localhost:6379/0
 
| `SERVICE_MESSAGE_QUEUE_REDIS`
| string
| ✔

> | URL of the job queue to schedule the indexing jobs for Redis.

Reviewer: ```suggestion
| URL of the Redis instance hosting the job queue to schedule the block indexing jobs.
```

| redis://localhost:6379/3

|===

--
Sample::
+
--
[source,js]
----
module.exports = {
  apps: [
    // Other types of configurations.
    {
      name: 'lisk-service-blockchain-coordinator',
      // [...]
      env: {
        // --- Remember to set the properties below
        SERVICE_BROKER: 'redis://localhost:6379/0',
        SERVICE_MESSAGE_QUEUE_REDIS: 'redis://localhost:6379/3',
      },
    },
    // Other types of configurations.
  ],
};
----
--
=====

=== Indexer

[tabs]
=====
Properties::
+
--
[cols="2,1,1,3,3",options="header",stripes="hover"]
|===

> |Property's Name
> |Datatype
> |Configurable

Reviewer: ```suggestion
|Environment variable
|Datatype
```

| string
| ✔
| Boolean flag to enable the Data Indexing mode.
| true

| `ENABLE_PERSIST_EVENTS`
| string
| ✔
| Boolean flag to permanently maintain the events in the MySQL database.
| false

| `SERVICE_INDEXER_CACHE_REDIS`
| string
| ✗
| URL of the cache storage (Redis).
| redis://localhost:6379/1
|===

--
Sample::
+
--
[source,js]
----
module.exports = {
  apps: [
    // Other types of configurations.
    {
      name: 'lisk-service-blockchain-indexer',
      // [...]
      env: {
        // --- Remember to set the properties below
        SERVICE_BROKER: 'redis://localhost:6379/0',
        SERVICE_INDEXER_CACHE_REDIS: 'redis://localhost:6379/1',
        SERVICE_INDEXER_REDIS_VOLATILE: 'redis://localhost:6379/2',
        SERVICE_MESSAGE_QUEUE_REDIS: 'redis://localhost:6379/3',
        SERVICE_INDEXER_MYSQL: 'mysql://lisk:password@localhost:3306/lisk',
        ENABLE_DATA_RETRIEVAL_MODE: 'true',
        ENABLE_INDEXING_MODE: 'true',
        ENABLE_PERSIST_EVENTS: 'false',
      },
    },
    // Other types of configurations.
  ],
};
----
--
=====

=== Blockchain application registry
[tabs]
=====
Properties::
+
--
[cols="2,1,1,3,3",options="header",stripes="hover"]
|===

> |Property's Name
> |Datatype
> |Configurable

Reviewer: ```suggestion
|Environment variable
|Datatype
```

|Description
|Sample

| `SERVICE_BROKER`
| string
| ✔
| URL of the microservice message broker (NATS or Redis).
| redis://localhost:6379/0
 
| `SERVICE_APP_REGISTRY_MYSQL`
| string
| ✔
| Connection string of the MySQL instance that the microservice connects to.
| mysql://lisk:password@localhost:3306/lisk

| `ENABLE_REBUILD_INDEX_AT_INIT`
| string
| ✔
| Boolean flag to truncate the index and rebuild at application init.
| false

|===

--
Sample::
+
--
[source,js]
----
module.exports = {
  apps: [
    // Other types of configurations.
    {
      name: 'lisk-service-blockchain-app-registry',
      // [...]
      env: {
        // --- Remember to set the properties below
        SERVICE_BROKER: 'redis://localhost:6379/0',
        SERVICE_APP_REGISTRY_MYSQL: 'mysql://lisk:password@localhost:3306/lisk',
        ENABLE_REBUILD_INDEX_AT_INIT: 'false',
      },
    },
    // Other types of configurations.
  ],
};
----
--
=====

=== Fee estimator

[tabs]
=====
Properties::
+
--
[cols="2,1,1,3,3",options="header",stripes="hover"]
|===

> |Property's Name
> |Datatype
> |Configurable

Reviewer: ```suggestion
|Environment variable
|Datatype
```

|Description
|Sample

| `SERVICE_BROKER`
| string
| ✔
| URL of the microservice message broker (NATS or Redis).
| redis://localhost:6379/0
 
| `SERVICE_FEE_ESTIMATOR_CACHE`
| string
| ✔
| URL of the cache storage (Redis).
| redis://localhost:6379/1

| `ENABLE_FEE_ESTIMATOR_QUICK`
| string
| ✗
| Enable quick algorithm for fee estimation.
| true

| `ENABLE_FEE_ESTIMATOR_FULL`
| string
| ✗
| Enable full algorithm for fee estimation.
| false

| `FEE_EST_COLD_START_BATCH_SIZE`
| string
| ✔
| Defines the number of blocks that are analyzed during cold start.
| 1

| `FEE_EST_DEFAULT_START_BLOCK_HEIGHT`
| string
| ✔

> | Defines the height at which the blockchain starts using dynamic fees.

Reviewer: ```suggestion
| Defines the block height at which the dynamic fee estimation algorithm starts.
```

| 1

|===

--
Sample::
+
--
[source,js]
----
module.exports = {
  apps: [
    // Other types of configurations.
    {
      name: 'lisk-service-fee-estimator',
      // [...]
      env: {
        // --- Remember to set the properties below
        SERVICE_BROKER: 'redis://localhost:6379/0',
        SERVICE_FEE_ESTIMATOR_CACHE: 'redis://localhost:6379/1',
        ENABLE_FEE_ESTIMATOR_QUICK: 'true',
        ENABLE_FEE_ESTIMATOR_FULL: 'false',
        FEE_EST_COLD_START_BATCH_SIZE: '1',
        FEE_EST_DEFAULT_START_BLOCK_HEIGHT: '1',
      },
    },
    // Other types of configurations.
  ],
};
----
--
=====

=== Transaction statistics

[tabs]
=====
Properties::
+
--
[cols="2,1,1,3,3",options="header",stripes="hover"]
|===

> |Property's Name
> |Datatype
> |Configurable

Reviewer: ```suggestion
|Environment variable
|Datatype
```

| redis://localhost:6379/0
 
| `SERVICE_STATISTICS_MYSQL`
| string
| ✔
| Connection string of the MySQL instance that the microservice connects to.
| mysql://lisk:password@localhost:3306/lisk

| `SERVICE_STATISTICS_REDIS`
| string
| ✔
| URL of the cache storage (Redis).
| redis://localhost:6379/1

| `TRANSACTION_STATS_HISTORY_LENGTH_DAYS`
| string
| ✔
| The number of days for which the transaction statistics need to be built in retrospect to the application init.
| 366

|===

--
Sample::
+
--
[source,js]
----
module.exports = {
  apps: [
    // Other types of configurations.
    {
      name: 'lisk-service-transaction-statistics',
      // [...]
      env: {
        // --- Remember to set the properties below
        SERVICE_BROKER: 'redis://localhost:6379/0',
        SERVICE_STATISTICS_REDIS: 'redis://localhost:6379/1',
        SERVICE_STATISTICS_MYSQL: 'mysql://lisk:password@localhost:3306/lisk',
        TRANSACTION_STATS_HISTORY_LENGTH_DAYS: '366',
      },
    },
    // Other types of configurations.
  ],
};
----
--
=====

=== Market

[tabs]
=====
Properties::
+
--
[cols="2,1,1,3,3",options="header",stripes="hover"]
|===

> |Property's Name
> |Datatype
> |Configurable

Reviewer: ```suggestion
|Environment variable
|Datatype
```

| `SERVICE_MARKET_TARGET_PAIRS`
| string
| ✗
| Exchange rates exposed to the Gateway.
The values listed here must be comma separated.
| LSK_BTC,LSK_EUR,BTC_CHF

| `EXCHANGERATESAPI_IO_API_KEY`
| string
| ✔
| Optional API key for https://exchangeratesapi.io/.
The `/market/prices` endpoint will respond with additional data, specifically the exchange rates for various cryptocurrencies in other fiat currencies.
The free plan would suffice for Lisk Service.
|  

|===

--
Sample::
+
--
[source,js]
----
module.exports = {
  apps: [
    // Other types of configurations.
    {
      name: 'lisk-service-market',
      // [...]
      env: {
        // --- Remember to set the properties below
        SERVICE_BROKER: 'redis://localhost:6379/0',
        SERVICE_MARKET_REDIS: 'redis://localhost:6379/2',
        SERVICE_MARKET_FIAT_CURRENCIES: 'EUR,USD,CHF,GBP,RUB',
        SERVICE_MARKET_TARGET_PAIRS: 'LSK_BTC,LSK_EUR,LSK_USD,LSK_CHF,BTC_EUR,BTC_USD,BTC_CHF',
        // EXCHANGERATESAPI_IO_API_KEY: ''
      },
    },
    // Other types of configurations.
  ],
};
----
--
=====

=== Export

[tabs]
=====
Properties::
+
--
[cols="2,1,1,3,3",options="header",stripes="hover"]
|===

> |Property's Name
> |Datatype
> |Configurable

Reviewer: ```suggestion
|Environment variable
|Datatype
```

|Description
|Sample

| `SERVICE_BROKER`
| string
| ✔
| URL of the microservice message broker (NATS or Redis).
| redis://localhost:6379/0
 
| `SERVICE_EXPORT_REDIS`
| string
| ✔
| URL of the permanent cache storage (Redis).
| redis://localhost:6379/3

| `SERVICE_EXPORT_REDIS_VOLATILE`
| string
| ✔
| URL of the volatile cache storage (Redis).
| redis://localhost:6379/4

|===

--
Sample::
+
--
[source,js]
----
module.exports = {
  apps: [
    // Other types of configurations.
    {
      name: 'lisk-service-export',
      // [...]
      env: {
        SERVICE_BROKER: 'redis://localhost:6379/0',
        SERVICE_EXPORT_REDIS: 'redis://localhost:6379/3',
        SERVICE_EXPORT_REDIS_VOLATILE: 'redis://localhost:6379/4',
      },
    },
    // Other types of configurations.
  ],
};
----
--
=====

== Lisk Settings

Configurable environment variables related to Lisk node.

=== Node settings

[cols="2,1,3,3",options="header",stripes="hover"]
|===
|Property's Name
|Datatype
|Description
|Sample

> | `LISK_CORE_WS`

Reviewer: We do not use this any more. It has been replaced by `LISK_APP_WS`

| string
| URL for Lisk Core WebSocket RPC API.
The URL will differ according to the environment i.e., docker, or server installation.
| ws://localhost:7887 or ws://host.docker.internal:7887
 

> | `LISK_CORE_CLIENT_TIMEOUT`

Reviewer: Same as above.

| string
| Lisk Core client timeout (in seconds).
| 30
|===

> === MySQL Snapshot settings for Docker

Reviewer: Will verify and update later.

Author: Will the verification be possible within this PR or should I create a new issue for this for a later time?

Reviewer: Let's create a separate PR for this.

Configurations to sync Lisk Service from an existing snapshot and to speed up the syncing process in the Docker environment.

[cols="2,1,3,3",options="header",stripes="hover"]
|===
|Property's Name
|Datatype
|Description
|Sample

| `ENABLE_APPLY_SNAPSHOT`
| string
| Enable or disable apply snapshot feature.
| true
 
| `INDEX_SNAPSHOT_URL`
| string
| Custom snapshot URL: expected to end with "sql.gz".
| https://snapshots.lisk.io/mainnet/service.sql.gz

| `DOCKER_COMPOSE_FILEPATH`
| string
| When MySQL is hosted as a docker-compose service, set the following environment variables.
Set docker-compose file path by using the absolute path.
| /Users/lisk/lisk-service/jenkins/mysql/docker-compose.yml

| `DOCKER_MYSQL_SERVICE_NAME`
| string
| Set MySQL service name as defined in the above docker-compose file.
| mysql

|===

== External services

[cols="2,1,3,3",options="header",stripes="hover"]
|===
|Property's Name
|Datatype
|Description
|Sample

> | `LISK_STATIC`

Reviewer: Currently, not in use.

Author: Should I remove this?

| string
| Lisk static assets, i.e., known account lists.
| https://static-data.lisk.com
 

> | `GEOIP_JSON`

Reviewer: Already mentioned within the connector. Do we need it here again?

Author: Can be removed!

= Lisk Service Configuration Reference
Muhammad Talha <muhammad.talha@lightcurve.io>
:toc:
:url_microservices: index.adoc#microservices

On the Introduction to Lisk Service page, we have discussed the various xref:{url_microservices}[microservices] that combine to form Lisk Service. On this page, we will thoroughly discuss the configuration options available for each service.

== Default configurations
By default, a Lisk Service installation will have an `ecosystem.config.js` file with all the default configurations.
The values of various environment variables are automatically picked by the microservice based on the installation environment.

> Other than that, many environment variables can be configured as per need, the details of which have been described with each microservice. 

Reviewer 2: ```suggestion
Aside from that, many environment variables can be configured to meet the specific requirements, and their details have been outlined with each microservice. 
```

				// EXCHANGERATESAPI_IO_API_KEY: ''
			},
		},
		{
			name: 'lisk-service-export',
			script: 'app.js',
			cwd: './services/export',
			pid_file: './pids/service_export.pid',
			out_file: './logs/service_export.log',
			error_file: './logs/service_export.err',
			log_date_format: 'YYYY-MM-DD HH:mm:ss SSS',
			watch: false,
			kill_timeout: 10000,
			max_memory_restart: '300M',
			autorestart: true,
			env: {
				SERVICE_BROKER: 'redis://localhost:6379/0',
				SERVICE_EXPORT_REDIS: 'redis://localhost:6379/3',
				SERVICE_EXPORT_REDIS_VOLATILE: 'redis://localhost:6379/4',
			},
		},
	],
};
----
====

== Microservice Configurations
In this section, we discuss the common and specific configurations of all the microservices of Lisk Service.

=== Common Configurations
The common configurations described in the section can be added to any microservice.
The value of each configuration defined below should be the same across all microservices.

[tabs]
=====
Properties::
+
--
[cols="2,1,1,3,3",options="header",stripes="hover"]
|===
|Property's Name
|Datatype
|Configurable
|Description
|Sample

| `SERVICE_BROKER`
| string
| ✔
| URL of the microservice message broker (NATS or Redis).

> Must be identical for all micro-services.
> Make sure that all micro-services can connect with the common Redis.

Author: ```suggestion
Must be identical for all microservices.
Make sure that all microservices can connect with the common Redis.
```

		{
			// Other types of configurations.
			{
			name: 'lisk-service-gateway',
			// [...]
			env: {
				// --- Remember to set the properties below
				SERVICE_BROKER: 'redis://localhost:6379/0',
				SERVICE_LOG_LEVEL: 'info',
				// [...]
			},
		},
		{
			name: 'lisk-service-transaction-statistics',
			// [...]
			env: {
				// --- Remember to set the properties below
				SERVICE_BROKER: 'redis://localhost:6379/0',
				SERVICE_LOG_LEVEL: 'info',
				// [...]
			},
		},
		// Other types of configurations.
	}
	],
};
----
--
=====

=== Gateway

[tabs]
=====
Properties::
+
--
[cols="2,1,1,3,3",options="header",stripes="hover"]
|===
|Property's Name
|Datatype
|Configurable
|Description
|Sample

| `PORT`
| string
| ✔
| Port that provides the possibility to connect with Lisk Service.
| 9901

| `HOST`
| string
| ✔
| URL of the host.
| 127.0.0.1
 
| `SERVICE_BROKER`
| string
| ✔
| URL of the microservice message broker (NATS or Redis).
| redis://localhost:6379/0

| `JSON_RPC_STRICT_MODE`
| string
| ✔
| Ensures that JSON-RPC accepts a valid JSON-RPC 2.0 envelope.
| false
 
| `ENABLE_HTTP_API`
| string
| ✔
| Enables particular HTTP APIs.
APIs must be listed as comma-separated values.

> | http-status,http-version3,http-exports

Reviewer 2: Fair point, then lets omit it. 👍

| `ENABLE_WS_API`
| string
| ✔
| Enables particular WebSocket APIs.
APIs must be listed as comma-separated values.

> | blockchain,rpc-v3

Reviewer 2: ```suggestion
| blockchain,rpc-v3.
```

Reviewer 2: As above.

> | blockchain,rpc-v3

Reviewer 2: ```suggestion
| blockchain,rpc-v3.
```

| `SERVICE_GATEWAY_REDIS_VOLATILE`
| string
| ✗
| URL of the volatile cache storage (Redis).
| redis://localhost:6379/5

| `GATEWAY_DEPENDENCIES`
| string
| ✗
| Describes the microservices on which the `gateway` service depends.

> | indexer,connector

Reviewer 2: ```suggestion
| indexer,connector.
```

|Configurable
|Description
|Sample

| `SERVICE_BROKER`
| string
| ✔
| URL of the microservice message broker (NATS or Redis).
| redis://localhost:6379/0
 
| `SERVICE_EXPORT_REDIS`
| string
| ✔
| URL of the permanent cache storage (Redis).
| redis://localhost:6379/3

| `SERVICE_EXPORT_REDIS_VOLATILE`
| string
| ✔
| URL of the volatile cache storage (Redis).
| redis://localhost:6379/4

|===

--
Sample::
+
--
[source,js]
----
module.exports = {
	apps: [
		{
            // Other types of configurations.
            {
                name: 'lisk-service-export',
                // [...]
                env: {
                    SERVICE_BROKER: 'redis://localhost:6379/0',
                    SERVICE_EXPORT_REDIS: 'redis://localhost:6379/3',
                    SERVICE_EXPORT_REDIS_VOLATILE: 'redis://localhost:6379/4',
                },
		    },
            // Other types of configurations.
        }
	],
};
----
--
=====

== Lisk Settings

Configurable environment variables related to Lisk node.

=== Node settings

[cols="2,1,3,3",options="header",stripes="hover"]
|===
|Property's Name
|Datatype
|Description
|Sample

| `LISK_CORE_WS`
| string
| URL for Lisk Core WebSocket RPC API.

> The URL will differ according to the environment i.e. docker, or server installation.

Reviewer 2: ```suggestion
The URL will differ according to the environment i.e., docker, or server installation.
```

| ws://localhost:7887 or ws://host.docker.internal:7887
 
| `LISK_CORE_CLIENT_TIMEOUT`
| string
| Lisk Core client timeout (in seconds).
| 30
|===

=== MySQL Snapshot settings for Docker
Configurations to sync Lisk Service from an existing snapshot and to speed up the syncing process in the Docker environment.

[cols="2,1,3,3",options="header",stripes="hover"]
|===
|Property's Name
|Datatype
|Description
|Sample

| `ENABLE_APPLY_SNAPSHOT`
| string
| Enable or disable apply snapshot feature.
| true
 
| `INDEX_SNAPSHOT_URL`
| string
| Custom snapshot URL: expected to end with "sql.gz".
| https://snapshots.lisk.io/mainnet/service.sql.gz

| `DOCKER_COMPOSE_FILEPATH`
| string
| When MySQL is hosted as a docker-compose service, set the following environment variables.
Set docker-compose file path by using the absolute path.
| /Users/lisk/lisk-service/jenkins/mysql/docker-compose.yml

| `DOCKER_MYSQL_SERVICE_NAME`
| string
| Set MySQL service name as defined in the above docker-compose file.
| mysql

|===

== External services

[cols="2,1,3,3",options="header",stripes="hover"]
|===
|Property's Name
|Datatype
|Description
|Sample

| `LISK_STATIC`
| string

> | Lisk static assets, i.e. known account lists.

Reviewer 2: ```suggestion
| Lisk static assets, i.e., known account lists.
```

  apps: [
    // Other types of configurations.
    {
      name: 'lisk-service-gateway',
      // [...]
      env: {
        // --- Remember to set the properties below
        SERVICE_BROKER: 'redis://localhost:6379/0',
        SERVICE_LOG_LEVEL: 'info',
        // [...]
      },
    },
    {
      name: 'lisk-service-transaction-statistics',
      // [...]
      env: {
        // --- Remember to set the properties below
        SERVICE_BROKER: 'redis://localhost:6379/0',
        SERVICE_LOG_LEVEL: 'info',
        // [...]
      },
    },
    // Other types of configurations.
  ],
};
----
--
=====

=== Gateway

[tabs]
=====
Properties::
+
--
[cols="2,1,3,3",options="header",stripes="hover"]
|===
|Environment variable
|Datatype
|Description
|Sample

| `PORT`
| string
| Port that provides the possibility to connect with Lisk Service.
| 9901

| `HOST`
| string
| URL of the host.
| 127.0.0.1
 
| `SERVICE_BROKER`
| string
| URL of the microservice message broker (NATS or Redis).
| redis://localhost:6379/0

| `JSON_RPC_STRICT_MODE`
| string
| Ensures that JSON-RPC accepts a valid JSON-RPC 2.0 envelope.
| false
 
| `ENABLE_HTTP_API`
| string
| Enables particular HTTP API endpoints.

> Directories (under `gateway/apis/`) containing the API method definitions must be listed as comma-separated values.

Author: ```suggestion
Directories (under `gateway/apis/`) containing the API method definitions.
Must be listed as comma-separated values.
```

| http-status,http-version3,http-exports

| `ENABLE_WS_API`
| string

> | Enables specified WebSocket API namespaces.
> APIs must be listed as comma-separated values.

Author: ```suggestion
| Enables specified WebSocket API namespaces.
Must be listed as comma-separated values.
```

| blockchain,rpc-v3

| `SERVICE_GATEWAY_REDIS_VOLATILE`
| string
| URL of the volatile cache storage (Redis). Used to cache RPC requests. Required when `ENABLE_REQUEST_CACHING` is enabled.
| redis://localhost:6379/5

| `GATEWAY_DEPENDENCIES`
| string
| Describes the microservices on which the `gateway` service depends.
| indexer,connector

| `WS_RATE_LIMIT_ENABLE`
| string
| To enable the WebSocket rate limit, this environment variable is required to be `true`.
| false

| `WS_RATE_LIMIT_CONNECTIONS`
| number
| Once the rate limit is enabled, this variable contains the number of connections per second.
| 5

| `WS_RATE_LIMIT_DURATION`
| number
| Defines the duration (in seconds) for which the WS rate should be limited.
| 1

| `ENABLE_REQUEST_CACHING`
| string

> | To enable RPC response caching, this environment variable is required to be `true`.

Author: ```suggestion
| To enable RPC response caching, this environment variable is required to be `true`.
Requires `SERVICE_GATEWAY_REDIS_VOLATILE` to be set.
```

        JSON_RPC_STRICT_MODE: "false",
        HTTP_RATE_LIMIT_ENABLE: "false",
        HTTP_RATE_LIMIT_CONNECTIONS: 200,
        HTTP_RATE_LIMIT_WINDOW: 10, // in seconds
        HTTP_CACHE_CONTROL_DIRECTIVES: "public, max-age=10",
        ENABLE_HTTP_CACHE_CONTROL: "true",
      },
    },
    // Other types of configurations.
  ],
};
----
--
=====

=== Connector

[tabs]
=====
Properties::
+
--
[cols="2,1,3,3",options="header",stripes="hover"]
|===
| Environment variable
|Datatype
|Description
|Sample

| `SERVICE_BROKER`
| string
| URL of the microservice message broker (NATS or Redis).
| redis://localhost:6379/0
 
| `LISK_APP_WS`
| string
| URL to connect with the Lisk application node over WebSocket.
| ws://localhost:7887

| `USE_LISK_IPC_CLIENT`
| string
| Boolean flag to enable IPC-based connection to the Lisk application node.
Requires `LISK_APP_DATA_PATH` to be set.
| true 

| `LISK_APP_DATA_PATH`
| string
| Data path to connect with the Lisk application node over IPC.
| ~/.lisk/lisk-core

| `GEOIP_JSON`
| string
| URL of the GeoIP server
| https://geoip.lisk.com/json

| `GENESIS_BLOCK_URL`
| string
| URL of the Lisk SDK-based application's genesis block.

> Only to be used when the genesis block is large enough to be transmitted over API calls within the timeout

Reviewer: Could we please ensure that all the descriptions end with a `.` for consistency?
```suggestion
Only to be used when the genesis block is large enough to be transmitted over API calls within the timeout.
```

| https://downloads.lisk.com/lisk/mainnet/genesis_block.json.tar.gz

|===

--
Sample::
+
--
[source,js]
----
module.exports = {
  apps: [
    // Other types of configurations.
    {
      name: "lisk-service-blockchain-connector",
      // [...]
      env: {
        // --- Remember to set the properties below
        SERVICE_BROKER: "redis://localhost:6379/0",
        LISK_APP_WS: "ws://localhost:7887",
        GEOIP_JSON: "https://geoip.lisk.com/json",
        USE_LISK_IPC_CLIENT: "true",
        GENESIS_BLOCK_URL: "",
        LISK_APP_DATA_PATH: "~/.lisk/lisk-core",
      },
    },
    // Other types of configurations.
  ],
};
----
--
=====

=== Coordinator

[tabs]
=====
Properties::
+
--
[cols="2,1,3,3",options="header",stripes="hover"]
|===
|Environment variable
|Datatype
|Description
|Sample

| `SERVICE_BROKER`
| string
| URL of the microservice message broker (NATS or Redis).
| redis://localhost:6379/0
 
| `SERVICE_MESSAGE_QUEUE_REDIS`
| string

> | URL of the Redis instance hosting the job queue to schedule the block indexing jobs.

Author: ```suggestion
| URL of the Redis instance hosting the job queue to schedule the block indexing jobs.
Must match the value supplied for the `indexer` microservice.
```

| redis://localhost:6379/3

|===

--
Sample::
+
--
[source,js]
----
module.exports = {
  apps: [
    // Other types of configurations.
    {
      name: 'lisk-service-blockchain-coordinator',
      // [...]
      env: {
        // --- Remember to set the properties below
        SERVICE_BROKER: 'redis://localhost:6379/0',
        SERVICE_MESSAGE_QUEUE_REDIS: 'redis://localhost:6379/3',
      },
    },
    // Other types of configurations.
  ],
};
----
--
=====

=== Indexer

[tabs]
=====
Properties::
+
--
[cols="2,1,3,3",options="header",stripes="hover"]
|===
|Environment variable
|Datatype
|Description
|Sample

| `SERVICE_BROKER`
| string
| URL of the microservice message broker (NATS or Redis).
| redis://localhost:6379/0
 
| `SERVICE_INDEXER_MYSQL`
| string
| Connection string of the MySQL instance that the microservice connects to.
| mysql://lisk:password@localhost:3306/lisk

| `SERVICE_MESSAGE_QUEUE_REDIS`
| string

> | URL of the job queue to process the scheduled indexing jobs by the Blockchain Coordinator (Redis).

Author: ```suggestion
| URL of the job queue to process the scheduled indexing jobs by the Blockchain Coordinator (Redis).
Must match the value supplied for the `coordinator` microservice.
```

        HTTP_RATE_LIMIT_WINDOW: 10, // in seconds
        HTTP_CACHE_CONTROL_DIRECTIVES: "public, max-age=10",
        ENABLE_HTTP_CACHE_CONTROL: "true",
      },
    },
    // Other types of configurations.
  ],
};
----
--
=====

=== Connector

[tabs]
=====
Properties::
+
--
[cols="2,1,3,3",options="header",stripes="hover"]
|===
| Environment variable
|Datatype
|Description
|Sample

| `SERVICE_BROKER`
| string
| URL of the microservice message broker (NATS or Redis).
| redis://localhost:6379/0
 
| `LISK_APP_WS`
| string
| URL to connect with the Lisk SDK-based application node over WebSocket.
| ws://localhost:7887

| `USE_LISK_IPC_CLIENT`
| string
| Boolean flag to enable IPC-based connection to the Lisk SDK-based application node.
Requires `LISK_APP_DATA_PATH` to be set.
| true 

| `LISK_APP_DATA_PATH`
| string
| Data path to connect with the Lisk SDK-based application node over IPC.
| ~/.lisk/lisk-core

| `GEOIP_JSON`
| string
| URL of the GeoIP server
| https://geoip.lisk.com/json

| `GENESIS_BLOCK_URL`
| string
| URL of the Lisk SDK-based application's genesis block.
Only to be used when the genesis block is large enough to be transmitted over API calls within the timeout.

> | 

Author: @sameersubudhi Please suggest a sample value for this environment variable.

Reviewer: `'https://downloads.lisk.com/lisk/mainnet/genesis_block.json.tar.gz'`

= Lisk Service Configuration Reference
Muhammad Talha <muhammad.talha@lightcurve.io>
:toc:
:url_microservices: index.adoc#microservices
:url_intro_services: index.adoc
:url_pm2: https://pm2.keymetrics.io/
:url_config_file: https://github.com/LiskHQ/lisk-service/blob/development/ecosystem.config.js
:url_lisk_service_repo: https://github.com/LiskHQ/lisk-service

> On the xref:{url_intro_services}[introduction to Lisk Service] page, we have discussed the various xref:{url_microservices}[microservices] that combine to form Lisk Service.

Reviewer: Assuming it's the title of the referred page.
```suggestion
On the xref:{url_intro_services}[Introduction to Lisk Service] page, we have discussed the various xref:{url_microservices}[microservices] that combine to form Lisk Service.
```

NOTE: The actual {url_config_file}[`*ecosystem.config.js*`] file for Lisk Service can be found in the {url_lisk_service_repo}[lisk-service] GitHub repository.

The values of various environment variables are automatically picked by the microservices based on the installation environment.
Aside from that, many environment variables can be configured to meet the specific requirements, the details of which have been outlined with each microservice.

== Microservice configurations
In this section, we discuss the common and specific configurations of all the microservices of Lisk Service.

=== Common configurations
The common configurations described in this section apply to all the microservices within Lisk Service.

[tabs]
=====
Properties::
+
--
[cols="2,1,3,3",options="header",stripes="hover"]
|===
|Environment variable
|Datatype
|Description
|Sample

| `SERVICE_BROKER`
| string
| URL of the microservice message broker.
*Must be identical for all microservices.*
Make sure that all microservices can connect with the message broker.

> This could be a NATS or Redis instance.

Reviewer: ```suggestion
This could be a NATS or a Redis instance.
```
